StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1214ReadonlyElementsMustAppearBeforeNonReadonlyElements.cs

```diff
@@ -0,0 +1,150 @@
+
+        private static readonly ImmutableArray<SyntaxKind> TypeDeclarationKinds =
+            ImmutableArray.Create(SyntaxKind.ClassDeclaration, SyntaxKind.StructDeclaration);
+
+        private static readonly Action<CompilationStartAnalysisContext> CompilationStartAction = HandleCompilationStart;
+        private static readonly Action<SyntaxNodeAnalysisContext, StyleCopSettings> TypeDeclarationAction = HandleTypeDeclaration;
+
+        /// <inheritdoc/>
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+            ImmutableArray.Create(Descriptor);
+
+        /// <inheritdoc/>
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(CompilationStartAction);
+        }
+
+        private static void HandleCompilationStart(CompilationStartAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionHonorExclusions(TypeDeclarationAction, TypeDeclarationKinds);
+        }
+
+        private static void HandleTypeDeclaration(SyntaxNodeAnalysisContext context, StyleCopSettings settings)
+        {
+            var elementOrder = settings.OrderingRules.ElementOrder;
+            int readonlyIndex = elementOrder.IndexOf(OrderingTrait.Readonly);
+            if (readonlyIndex < 0)
+            {
+                return;
+            }
+
+            var typeDeclaration = (TypeDeclarationSyntax)context.Node;
+
+            FieldDeclarationSyntax previousMember = null;
+            var previousFieldConst = true;
+            var previousFieldStatic = false;
+            var previousFieldReadonly = false;
+            var previousAccessLevel = AccessLevel.NotSpecified;
+            foreach (var member in typeDeclaration.Members)
+            {
+                FieldDeclarationSyntax field = member as FieldDeclarationSyntax;
+                if (field == null)
+                {
+                    previousMember = null;
+                    continue;
+                }
+
+                var modifiers = member.GetModifiers();
+                var currentAccessLevel = MemberOrderHelper.GetAccessLevelForOrdering(member, modifiers);
+                bool currentFieldConst = modifiers.Any(SyntaxKind.ConstKeyword);
+                bool currentFieldStatic = modifiers.Any(SyntaxKind.StaticKeyword);
+                bool currentFieldReadonly = modifiers.Any(SyntaxKind.ReadOnlyKeyword);
+                if (previousMember == null)
+                {
+                    previousMember = field;
+                    previousFieldConst = currentFieldConst;
+                    previousFieldStatic = currentFieldStatic;
+                    previousFieldReadonly = currentFieldReadonly;
+                    previousAccessLevel = currentAccessLevel;
+                    continue;
+                }
+
+                bool compareReadonly = true;
+                for (int j = 0; j < readonlyIndex; j++)
```

Comment:
:thought_balloon: consider adding `compareReadonly` to the condition check, to stop parsing the element order as soon as the outcome is known.